src/web/routes.rs

```diff
@@ -0,0 +1,257 @@
+    );
+    routes.internal_page(
+        "/releases/failures",
+        super::releases::releases_failures_by_stars_handler,
+    );
+    routes.internal_page(
+        "/releases/failures/:page",
+        super::releases::releases_failures_by_stars_handler,
+    );
+
+    routes.internal_page("/crate/:name", super::crate_details::crate_details_handler);
+    routes.internal_page(
+        "/crate/:name/:version",
+        super::crate_details::crate_details_handler,
+    );
+    routes.internal_page(
+        "/crate/:name/:version/builds",
+        super::builds::build_list_handler,
+    );
+    routes.static_resource(
+        "/crate/:name/:version/builds.json",
+        super::builds::build_list_handler,
+    );
+    routes.internal_page(
+        "/crate/:name/:version/builds/:id",
+        super::builds::build_list_handler,
+    );
+    routes.internal_page(
+        "/crate/:name/:version/source/",
+        super::source::source_browser_handler,
+    );
+    routes.internal_page(
+        "/crate/:name/:version/source/*",
+        super::source::source_browser_handler,
+    );
+
+    routes.rustdoc_page("/:crate", super::rustdoc::rustdoc_redirector_handler);
+    routes.rustdoc_page("/:crate/", super::rustdoc::rustdoc_redirector_handler);
+    routes.rustdoc_page("/:crate/badge.svg", super::rustdoc::badge_handler);
+    routes.rustdoc_page(
+        "/:crate/:version",
+        super::rustdoc::rustdoc_redirector_handler,
+    );
+    routes.rustdoc_page(
+        "/:crate/:version/",
+        super::rustdoc::rustdoc_redirector_handler,
+    );
+    routes.rustdoc_page(
+        "/:crate/:version/settings.html",
+        super::rustdoc::rustdoc_html_server_handler,
+    );
+    routes.rustdoc_page(
+        "/:crate/:version/all.html",
+        super::rustdoc::rustdoc_html_server_handler,
+    );
+    routes.rustdoc_page(
+        "/:crate/:version/:target",
+        super::rustdoc::rustdoc_redirector_handler,
+    );
+    routes.rustdoc_page(
+        "/:crate/:version/:target/",
+        super::rustdoc::rustdoc_html_server_handler,
+    );
+    routes.rustdoc_page(
+        "/:crate/:version/:target/*.html",
+        super::rustdoc::rustdoc_html_server_handler,
+    );
+
+    for redirect in DOC_RUST_LANG_ORG_REDIRECTS {
+        routes.internal_page(
+            &format!("/{}", redirect),
+            super::rustdoc::RustLangRedirector::new(redirect),
+        );
+    }
+
+    routes
+}
+
+pub(super) struct Routes {
+    get: Vec<(String, Box<dyn Handler>)>,
+    rustdoc_get: Vec<(String, Box<dyn Handler>)>,
+    page_prefixes: HashSet<String>,
+}
+
+impl Routes {
+    fn new() -> Self {
+        Self {
+            get: Vec::new(),
+            rustdoc_get: Vec::new(),
+            page_prefixes: HashSet::new(),
+        }
+    }
+
+    pub(super) fn page_prefixes(&self) -> HashSet<String> {
+        self.page_prefixes.clone()
+    }
+
+    pub(super) fn iron_router(mut self) -> Router {
```

Comment:
Oh I see, this whole `Routes` struct is a builder for `Router`. That would be nice to have in comments up at the top.